File: source/src/ECS/Core/ComponentStorage.ts
import { Component } from '../Component';

/**
 * 组件类型定义
 */
export type ComponentType<T extends Component = Component> = new (...args: any[]) => T;

/**
 * 组件注册表
 * 管理组件类型的位掩码分配
 */
export class ComponentRegistry {
    private static componentTypes = new Map<Function, number>();
    private static nextBitIndex = 0;
    private static maxComponents = 64; // 支持最多64种组件类型

    /**
     * 注册组件类型并分配位掩码
     * @param componentType 组件类型
     * @returns 分配的位索引
     */
    public static register<T extends Component>(componentType: ComponentType<T>): number {
        if (this.componentTypes.has(componentType)) {
            return this.componentTypes.get(componentType)!;
        }

        if (this.nextBitIndex >= this.maxComponents) {
            throw new Error(`Maximum number of component types (${this.maxComponents}) exceeded`);
        }

        const bitIndex = this.nextBitIndex++;
        this.componentTypes.set(componentType, bitIndex);
        return bitIndex;
    }

    /**
     * 获取组件类型的位掩码
     * @param componentType 组件类型
     * @returns 位掩码
     */
    public static getBitMask<T extends Component>(componentType: ComponentType<T>): bigint {
        const bitIndex = this.componentTypes.get(componentType);
        if (bitIndex === undefined) {
            throw new Error(`Component type ${componentType.name} is not registered`);
        }
        return BigInt(1) << BigInt(bitIndex);
    }

    /**
     * 获取组件类型的位索引
     * @param componentType 组件类型
     * @returns 位索引
     */
    public static getBitIndex<T extends Component>(componentType: ComponentType<T>): number {
        const bitIndex = this.componentTypes.get(componentType);
        if (bitIndex === undefined) {
            throw new Error(`Component type ${componentType.name} is not registered`);
        }
        return bitIndex;
    }

    /**
     * 检查组件类型是否已注册
     * @param componentType 组件类型
     * @returns 是否已注册
     */
    public static isRegistered<T extends Component>(componentType: ComponentType<T>): boolean {
        return this.componentTypes.has(componentType);
    }

    /**
     * 获取所有已注册的组件类型
     * @returns 组件类型映射
     */
    public static getAllRegisteredTypes(): Map<Function, number> {
        return new Map(this.componentTypes);
    }
}

/**
 * 高性能组件存储器
 * 使用SoA（Structure of Arrays）模式存储组件
 */
export class ComponentStorage<T extends Component> {
    private components: (T | null)[] = [];
    private entityToIndex = new Map<number, number>();
    private indexToEntity: number[] = [];
    private freeIndices: number[] = [];
    private componentType: ComponentType<T>;
    private _size = 0;

    constructor(componentType: ComponentType<T>) {
        this.componentType = componentType;
        
        // 确保组件类型已注册
        if (!ComponentRegistry.isRegistered(componentType)) {
            ComponentRegistry.register(componentType);
        }
    }

    /**
     * 添加组件
     * @param entityId 实体ID
     * @param component 组件实例
     */
    public addComponent(entityId: number, component: T): void {
        // 检查实体是否已有此组件
        if (this.entityToIndex.has(entityId)) {
            throw new Error(`Entity ${entityId} already has component ${this.componentType.name}`);
        }

        let index: number;
        
        if (this.freeIndices.length > 0) {
            // 重用空闲索引
            index = this.freeIndices.pop()!;
            this.components[index] = component;
            this.indexToEntity[index] = entityId;
        } else {
            // 添加到末尾
            index = this.components.length;
            this.components.push(component);
            this.indexToEntity.push(entityId);
        }
        
        this.entityToIndex.set(entityId, index);
        this._size++;
    }

    /**
     * 获取组件
     * @param entityId 实体ID
     * @returns 组件实例或null
     */
    public getComponent(entityId: number): T | null {
        const index = this.entityToIndex.get(entityId);
        return index !== undefined ? this.components[index] : null;
    }

    /**
     * 检查实体是否有此组件
     * @param entityId 实体ID
     * @returns 是否有组件
     */
    public hasComponent(entityId: number): boolean {
        return this.entityToIndex.has(entityId);
    }

    /**
     * 移除组件
     * @param entityId 实体ID
     * @returns 被移除的组件或null
     */
    public removeComponent(entityId: number): T | null {
        const index = this.entityToIndex.get(entityId);
        if (index === undefined) {
            return null;
        }

        const component = this.components[index];
        this.entityToIndex.delete(entityId);
        this.components[index] = null;
        this.freeIndices.push(index);
        this._size--;

        return component;
    }

    /**
     * 高效遍历所有组件
     * @param callback 回调函数
     */
    public forEach(callback: (component: T, entityId: number, index: number) => void): void {
        for (let i = 0; i < this.components.length; i++) {
            const component = this.components[i];
            if (component) {
                callback(component, this.indexToEntity[i], i);
            }
        }
    }

    /**
     * 获取所有组件（密集数组）
     * @returns 组件数组
     */
    public getDenseArray(): { components: T[]; entityIds: number[] } {
        const components: T[] = [];
        const entityIds: number[] = [];

        for (let i = 0; i < this.components.length; i++) {
            const component = this.components[i];
            if (component) {
                components.push(component);
                entityIds.push(this.indexToEntity[i]);
            }
        }

        return { components, entityIds };
    }

    /**
     * 清空所有组件
     */
    public clear(): void {
        this.components.length = 0;
        this.entityToIndex.clear();
        this.indexToEntity.length = 0;
        this.freeIndices.length = 0;
        this._size = 0;
    }

    /**
     * 获取组件数量
     */
    public get size(): number {
        return this._size;
    }

    /**
     * 获取组件类型
     */
    public get type(): ComponentType<T> {
        return this.componentType;
    }

    /**
     * 压缩存储（移除空洞）
     */
    public compact(): void {
        if (this.freeIndices.length === 0) {
            return; // 没有空洞，无需压缩
        }

        const newComponents: T[] = [];
        const newIndexToEntity: number[] = [];
        const newEntityToIndex = new Map<number, number>();

        let newIndex = 0;
        for (let i = 0; i < this.components.length; i++) {
            const component = this.components[i];
            if (component) {
                newComponents[newIndex] = component;
                newIndexToEntity[newIndex] = this.indexToEntity[i];
                newEntityToIndex.set(this.indexToEntity[i], newIndex);
                newIndex++;
            }
        }

        this.components = newComponents;
        this.indexToEntity = newIndexToEntity;
        this.entityToIndex = newEntityToIndex;
        this.freeIndices.length = 0;
    }

    /**
     * 获取存储统计信息
     */
    public getStats(): {
        totalSlots: number;
        usedSlots: number;
        freeSlots: number;
        fragmentation: number;
    } {
        const totalSlots = this.components.length;
        const usedSlots = this._size;
        const freeSlots = this.freeIndices.length;
        const fragmentation = totalSlots > 0 ? freeSlots / totalSlots : 0;

        return {
            totalSlots,
            usedSlots,
            freeSlots,
            fragmentation
        };
    }
}

/**
 * 组件存储管理器
 * 管理所有组件类型的存储器
 */
export class ComponentStorageManager {
    private storages = new Map<Function, ComponentStorage<any>>();

    /**
     * 获取或创建组件存储器
     * @param componentType 组件类型
     * @returns 组件存储器
     */
    public getStorage<T extends Component>(componentType: ComponentType<T>): ComponentStorage<T> {
        let storage = this.storages.get(componentType);
        
        if (!storage) {
            storage = new ComponentStorage(componentType);
            this.storages.set(componentType, storage);
        }
        
        return storage;
    }

    /**
     * 添加组件
     * @param entityId 实体ID
     * @param component 组件实例
     */
    public addComponent<T extends Component>(entityId: number, component: T): void {
        const componentType = component.constructor as ComponentType<T>;
        const storage = this.getStorage(componentType);
        storage.addComponent(entityId, component);
    }

    /**
     * 获取组件
     * @param entityId 实体ID
     * @param componentType 组件类型
     * @returns 组件实例或null
     */
    public getComponent<T extends Component>(entityId: number, componentType: ComponentType<T>): T | null {
        const storage = this.storages.get(componentType);
        return storage ? storage.getComponent(entityId) : null;
    }

    /**
     * 检查实体是否有组件
     * @param entityId 实体ID
     * @param componentType 组件类型
     * @returns 是否有组件
     */
    public hasComponent<T extends Component>(entityId: number, componentType: ComponentType<T>): boolean {
        const storage = this.storages.get(componentType);
        return storage ? storage.hasComponent(entityId) : false;
    }

    /**
     * 移除组件
     * @param entityId 实体ID
     * @param componentType 组件类型
     * @returns 被移除的组件或null
     */
    public removeComponent<T extends Component>(entityId: number, componentType: ComponentType<T>): T | null {
        const storage = this.storages.get(componentType);
        return storage ? storage.removeComponent(entityId) : null;
    }

    /**
     * 移除实体的所有组件
     * @param entityId 实体ID
     */
    public removeAllComponents(entityId: number): void {
        for (const storage of this.storages.values()) {
            storage.removeComponent(entityId);
        }
    }

    /**
     * 获取实体的组件位掩码
     * @param entityId 实体ID
     * @returns 组件位掩码
     */
    public getComponentMask(entityId: number): bigint {
        let mask = BigInt(0);
        
        for (const [componentType, storage] of this.storages.entries()) {
            if (storage.hasComponent(entityId)) {
                mask |= ComponentRegistry.getBitMask(componentType as ComponentType);
            }
        }
        
        return mask;
    }

    /**
     * 压缩所有存储器
     */
    public compactAll(): void {
        for (const storage of this.storages.values()) {
            storage.compact();
        }
    }

    /**
     * 获取所有存储器的统计信息
     */
    public getAllStats(): Map<string, any> {
        const stats = new Map<string, any>();
        
        for (const [componentType, storage] of this.storages.entries()) {
            const typeName = (componentType as any).name || 'Unknown';
            stats.set(typeName, storage.getStats());
        }
        
        return stats;
    }

    /**
     * 清空所有存储器
     */
    public clear(): void {
        for (const storage of this.storages.values()) {
            storage.clear();
        }
        this.storages.clear();
    }
}

File: packages/core-rust/src/storage/component_storage.rs
use crate::core::component::{Component, ComponentRegistry};
use rustc_hash::FxHashMap;
use std::any::TypeId;

/**
 * 高性能组件存储器
 */
pub struct ComponentStorage<T: Component + 'static> {
    components: Vec<Option<T>>,
    entity_to_index: FxHashMap<u32, usize>,
    index_to_entity: Vec<u32>,
    free_indices: Vec<usize>,
    size: usize,
}

impl<T: Component + 'static> ComponentStorage<T> {
    pub fn new() -> Self {
        Self {
            components: Vec::new(),
            entity_to_index: FxHashMap::default(),
            index_to_entity: Vec::new(),
            free_indices: Vec::new(),
            size: 0,
        }
    }

    /**
     * 添加组件
     * @param entity_id 实体ID
     * @param component 组件实例
     */
    pub fn add_component(&mut self, entity_id: u32, component: T) -> Result<(), String> {
        if self.entity_to_index.contains_key(&entity_id) {
            return Err(format!("Entity {} already has this component", entity_id));
        }

        let index = if let Some(free_index) = self.free_indices.pop() {
            self.components[free_index] = Some(component);
            self.index_to_entity[free_index] = entity_id;
            free_index
        } else {
            let index = self.components.len();
            self.components.push(Some(component));
            self.index_to_entity.push(entity_id);
            index
        };

        self.entity_to_index.insert(entity_id, index);
        self.size += 1;
        Ok(())
    }

    /**
     * 获取组件
     * @param entity_id 实体ID
     * @returns 组件引用或None
     */
    pub fn get_component(&self, entity_id: u32) -> Option<&T> {
        self.entity_to_index
            .get(&entity_id)
            .and_then(|&index| self.components.get(index))
            .and_then(|component| component.as_ref())
    }

    /**
     * 获取可变组件
     * @param entity_id 实体ID
     * @returns 可变组件引用或None
     */
    pub fn get_component_mut(&mut self, entity_id: u32) -> Option<&mut T> {
        self.entity_to_index
            .get(&entity_id)
            .and_then(|&index| self.components.get_mut(index))
            .and_then(|component| component.as_mut())
    }

    /**
     * 检查实体是否有此组件
     * @param entity_id 实体ID
     * @returns 是否有组件
     */
    pub fn has_component(&self, entity_id: u32) -> bool {
        self.entity_to_index.contains_key(&entity_id)
    }

    /**
     * 移除组件
     * @param entity_id 实体ID
     * @returns 被移除的组件或None
     */
    pub fn remove_component(&mut self, entity_id: u32) -> Option<T> {
        let index = self.entity_to_index.remove(&entity_id)?;
        let component = self.components[index].take()?;
        self.free_indices.push(index);
        self.size -= 1;
        Some(component)
    }

    /**
     * 高效遍历所有组件
     * @param callback 回调函数
     */
    pub fn for_each<F>(&self, mut callback: F)
    where
        F: FnMut(&T, u32, usize),
    {
        for (index, component) in self.components.iter().enumerate() {
            if let Some(comp) = component {
                let entity_id = self.index_to_entity[index];
                callback(comp, entity_id, index);
            }
        }
    }

    /**
     * 获取所有组件（密集数组）
     */
    pub fn get_dense_array(&self) -> (Vec<&T>, Vec<u32>) {
        let mut components = Vec::new();
        let mut entity_ids = Vec::new();

        for (index, component) in self.components.iter().enumerate() {
            if let Some(comp) = component {
                components.push(comp);
                entity_ids.push(self.index_to_entity[index]);
            }
        }

        (components, entity_ids)
    }

    /**
     * 清空所有组件
     */
    pub fn clear(&mut self) {
        self.components.clear();
        self.entity_to_index.clear();
        self.index_to_entity.clear();
        self.free_indices.clear();
        self.size = 0;
    }

    /**
     * 获取组件数量
     */
    pub fn size(&self) -> usize {
        self.size
    }

    /**
     * 获取存储统计信息
     */
    pub fn get_stats(&self) -> ComponentStorageStats {
        let total_slots = self.components.len();
        let used_slots = self.size;
        let free_slots = self.free_indices.len();
        let fragmentation = if total_slots > 0 {
            free_slots as f32 / total_slots as f32
        } else {
            0.0
        };

        ComponentStorageStats {
            total_slots,
            used_slots,
            free_slots,
            fragmentation,
        }
    }
}

/**
 * 组件存储统计信息
 */
pub struct ComponentStorageStats {
    pub total_slots: usize,
    pub used_slots: usize,
    pub free_slots: usize,
    pub fragmentation: f32,
}

/**
 * 组件存储管理器
 * 管理所有组件类型的存储器
 */
pub struct ComponentStorageManager {
    storages: FxHashMap<TypeId, Box<dyn ComponentStorageInterface>>,
    registry: ComponentRegistry,
}

impl ComponentStorageManager {
    pub fn new() -> Self {
        Self {
            storages: FxHashMap::default(),
            registry: ComponentRegistry::new(),
        }
    }

    /**
     * 获取组件注册表
     */
    pub fn get_registry(&self) -> &ComponentRegistry {
        &self.registry
    }

    /**
     * 获取可变组件注册表
     */
    pub fn get_registry_mut(&mut self) -> &mut ComponentRegistry {
        &mut self.registry
    }

    /**
     * 添加组件
     * @param entity_id 实体ID
     * @param component 组件实例
     */
    pub fn add_component<T: Component + Clone + 'static>(&mut self, entity_id: u32, component: T) -> Result<(), String> {
        let type_id = TypeId::of::<T>();
        
        if !self.storages.contains_key(&type_id) {
            let storage = Box::new(ComponentStorage::<T>::new());
            self.storages.insert(type_id, storage);
        }

        if let Some(storage) = self.storages.get_mut(&type_id) {
            storage.add_component_boxed(entity_id, Box::new(component))
        } else {
            Err("Failed to get storage".to_string())
        }
    }

    /**
     * 移除实体的所有组件
     * @param entity_id 实体ID
     */
    pub fn remove_all_components(&mut self, entity_id: u32) {
        for storage in self.storages.values_mut() {
            storage.remove_component_by_entity(entity_id);
        }
    }

    /**
     * 获取实体的组件位掩码
     * @param entity_id 实体ID
     * @returns 组件位掩码
     */
    pub fn get_component_mask(&self, entity_id: u32) -> u64 {
        let mut mask = 0u64;
        for (&type_id, storage) in &self.storages {
            if storage.has_component(entity_id) {
                if let Some(bit_mask) = self.registry.get_bit_mask(&type_id) {
                    mask |= bit_mask;
                }
            }
        }
        mask
    }

    /**
     * 清空所有存储器
     */
    pub fn clear(&mut self) {
        for storage in self.storages.values_mut() {
            storage.clear();
        }
        self.storages.clear();
    }
}

/**
 * 组件存储接口
 */
trait ComponentStorageInterface: Send + Sync {
    fn add_component_boxed(&mut self, entity_id: u32, component: Box<dyn Component>) -> Result<(), String>;
    fn remove_component_by_entity(&mut self, entity_id: u32) -> bool;
    fn has_component(&self, entity_id: u32) -> bool;
    fn clear(&mut self);
}

impl<T: Component + Clone + 'static> ComponentStorageInterface for ComponentStorage<T> {
    fn add_component_boxed(&mut self, entity_id: u32, component: Box<dyn Component>) -> Result<(), String> {
        if let Some(downcasted) = component.as_any().downcast_ref::<T>() {
            self.add_component(entity_id, downcasted.clone())
        } else {
            Err("Failed to downcast component".to_string())
        }
    }

    fn remove_component_by_entity(&mut self, entity_id: u32) -> bool {
        self.remove_component(entity_id).is_some()
    }

    fn has_component(&self, entity_id: u32) -> bool {
        ComponentStorage::has_component(self, entity_id)
    }

    fn clear(&mut self) {
        ComponentStorage::clear(self)
    }
}